spec/controllers/sys_controller_spec.rb

```diff
@@ -28,208 +28,368 @@
+          expect(response.code).to eq("403")
+        end
       end
-    end
 
-    describe '#repo_auth', 'for invalid api key' do
-      before(:each) do
-        @key = 'invalid'
+      context 'for valid login and project is public' do
+        before(:each) do
+          @key = Setting.sys_api_key
+          @project = FactoryGirl.create(:project, :is_public => true)
+
+          random_project = FactoryGirl.create(:project, :is_public => false)
+          @member = FactoryGirl.create(:member, :user => valid_user,
+                                                :roles => [browse_role],
+                                                :project => random_project)
+
+          request.env['HTTP_AUTHORIZATION'] = ActionController::HttpAuthentication::Basic.encode_credentials(valid_user.login, valid_user_password)
+          post "repo_auth", { :key => @key, :repository => @project.identifier, :method => "GET", :git_smart_http => '1', :uri => "/git", :location => '/git' }
+        end
+
+        it "should respond 200 OK" do
+          expect(response.code).to eq("200")
+        end
       end
 
-      it 'should respond 403 for valid username/password' do
-        request.env['HTTP_AUTHORIZATION'] = ActionController::HttpAuthentication::Basic.encode_credentials(valid_user.login, valid_user_password)
-        post 'repo_auth', key: @key, repository: 'any-repo', method: 'GET'
-        expect(response.code).to eq('403')
-        expect(response.body).to eq('Access denied. Repository management WS is disabled or key is invalid.')
+      context 'for invalid credentials' do
+        before(:each) do
+          @key = Setting.sys_api_key
+          post "repo_auth", { :key => @key, :repository => "any-repo", :method => "GET", :git_smart_http => '1', :uri => "/git", :location => '/git' }
+        end
+
+        it "should respond 401 auth required" do
+          expect(response.code).to eq("401")
+          expect(response.body).to eq("Authorization required")
+        end
       end
 
-      it 'should respond 403 for invalid username/password' do
-        request.env['HTTP_AUTHORIZATION'] = ActionController::HttpAuthentication::Basic.encode_credentials('invalid', 'invalid')
-        post 'repo_auth', key: @key, repository: 'any-repo', method: 'GET'
-        expect(response.code).to eq('403')
-        expect(response.body).to eq('Access denied. Repository management WS is disabled or key is invalid.')
+      context 'for invalid api key' do
+        before(:each) do
+          @key = "invalid"
+        end
+
+        it "should respond 403 for valid username/password" do
+          request.env['HTTP_AUTHORIZATION'] = ActionController::HttpAuthentication::Basic.encode_credentials(valid_user.login, valid_user_password)
+          post "repo_auth", { :key => @key, :repository => "any-repo", :method => "GET", :git_smart_http => '1', :uri => "/git", :location => '/git' }
+          expect(response.code).to eq("403")
```

Comment:
Prefer single-quoted strings when you don't need string interpolation or special symbols.